### ISSUE TITLE
Host key included during connection when I set postgres host config to blank

### ISSUE THREAD

**Issue by aydot** [cakephp/cakephp#6368]
According to the docs [http://book.cakephp.org/3.0/en/orm/database-basics.html#configuration]
```
unix_socket
    Used by drivers that support it to connect via Unix socket files. If you are using PostgreSQL and want to use Unix sockets, leave the host key blank.
```
However, when I leave the host key as blank, I get this error:
```
SQLSTATE[08006] [7] could not translate host name "port=5432" to address: Name or service not known 
```
My intention was to leave out the host param when connecting - according to 
[http://php.net/manual/en/ref.pdo-pgsql.connection.php#80091]

**Comment by markstory** [comment]
Are you defining port when using the unix_socket?

**Comment by aydot** [comment]
On properly configured server, this works -
```
$dsn = "pgsql:dbname={$config['database']}";
```
So didn't really need to define the port

**Comment by oldskool** [comment]
Did you configure your database with a config array or a DSN?
When using this config, everything seems to work as expected on my end:

```php
$config = [
    'persistent' => false,
    'host' => '',
    'database' => 'cake',
    'username' => 'user',
    'password' => 'pass',
    'port' => 5432,
    'flags' => [1 => true, 2 => false],
    'encoding' => 'a-language',
    'timezone' => 'Antartica',
    'schema' => 'fooblic',
    'unix_socket' => true,
    'init' => ['Execute this', 'this too']
];
```

**Comment by 1worldtech-bhoffman** [comment]
Hello,
When building the DSN here : http://api.cakephp.org/3.0/source-class-Cake.Database.Driver.Postgres.html#61

I believe that as soon as you define 'host=' in the DSN it will attempt to connect over TCP and not directly to the unix socket.  

Some examples:
All of these tests have been performed after disabling TCP communication on localhost over default port on linux server and postgres where linux user and postgres user exists.  

Connects properly over socket
```php
// config/app.php
'Datasources' => [
        'default' => [
            'className' => 'Cake\Database\Connection',
            'driver' => 'Cake\Database\Driver\Postgres',
            'database' => 'cake',
            'username' => 'user',
     ]
],
// \Cake\Database\Driver\Postgres
$dsn = "pgsql:dbname={$config['database']}";
```


Correctly fails when not defining the user with error: Error: SQLSTATE[08006] [7] FATAL: Peer authentication failed for user "root" 
```php
// config/app.php
'Datasources' => [
        'default' => [
            'className' => 'Cake\Database\Connection',
            'driver' => 'Cake\Database\Driver\Postgres',
            'database' => 'cake',
            // 'username' => 'user',
     ]
],

// \Cake\Database\Driver\Postgres
$dsn = "pgsql:dbname={$config['database']}";
```

Connection times out : Error: SQLSTATE[08006] [7] timeout expired 
This is expected as the DSN as defined is attempting to connect to localhost over TCP which has been disabled on test server via linux firewall 
```php
// config/app.php
'Datasources' => [
        'default' => [
            'className' => 'Cake\Database\Connection',
            'driver' => 'Cake\Database\Driver\Postgres',
            'database' => 'cake',
            'username' => 'user',

            'host' => 'localhost',
     ]
],

// \Cake\Database\Driver\Postgres
$dsn = "pgsql:host={$config['host']};port={$config['port']};dbname={$config['database']}";
```

Fails with error:  Error: SQLSTATE[08006] [7] could not translate host name "port=5432" to address: Name or service not known 
```php
// config/app.php
'Datasources' => [
        'default' => [
            'className' => 'Cake\Database\Connection',
            'driver' => 'Cake\Database\Driver\Postgres',
            'database' => 'cake',
            'username' => 'user',

            'host' => '',
            'unix_socket' => true, // as suggested above by oldskool, with or without this setting the same error occurs
     ]
],

// \Cake\Database\Driver\Postgres
$dsn = "pgsql:host={$config['host']};port={$config['port']};dbname={$config['database']}";
```


Per the docs:  http://book.cakephp.org/3.0/en/orm/database-basics.html#configuration

If I "leave the host key blank"  the default $_baseConfig will write 'host' => 'localhost' by default
http://api.cakephp.org/3.0/source-class-Cake.Database.Driver.Postgres.html#31

**Comment by markstory** [comment]
Thanks for the additional information @1worldtech-bhoffman I can take a look at how we could prioritize the unix_socket mode other host/port.

**Comment by markstory** [comment]
Fixed with #6438

**Comment by bhoffman13** [comment]
@markstory I believe I led us to an incorrect/inconsistent solution.  
According to the docs the solution implemented will work:  http://www.postgresql.org/docs/8.3/static/libpq-connect.html
"Without either a host name or host address, libpq will connect using a local Unix-domain socket; or on machines without Unix-domain sockets, it will attempt to connect to localhost."

However when using plugins or 3rd party libraries such as Migrations Plugin we discovered that the solution implemented causes issues as phinx handles the connection to postgres similar to CakePHP did prior to the fix.  

I think the correct fix is a matter of proper documentation and proper permissions to the database server.
I found this page helpful in correcting the issue in our environments: https://help.ubuntu.com/community/PostgreSQL

Using a DSN similar to this:  ```" psql "host=/tmp port=5433 user=bob ..."``` where the socket directory is specified in the host value works properly for common usage of PDO_PGSQL DSN (i.e. phinx https://github.com/robmorgan/phinx/blob/0.4.x-dev/src/Phinx/Db/Adapter/PostgresAdapter.php#L62)

The issue with using the unix_socket_directories is that you need heightened permissions on the database server to run the following command to discover where the socket directory resides 
```sudo -u postgres psql -c "SHOW unix_socket_directories;"```  i.e. in our case the socket directory was /var/run/postgresql

In summary I think the docs should provide details on how to create the DB connection and point developers in the right direction to the documentation needed to create the proper connection to the unix socket.

a sample app.php datasource
```
    'Datasources' => [
        'default' => [
            'className' => 'Cake\Database\Connection',
            'driver' => 'Cake\Database\Driver\Postgres',
            'persistent' => false, 
            'host' => '/var/run/postgresql',    // in pdo it will use unix socket
            'username' => 'user',
            'password' => 'password',
            'database' => 'db_name',
            'encoding' => 'utf8',
            'timezone' => 'UTC',
            'cacheMetadata' => true,
            'quoteIdentifiers' => false,
        ],
```

**Comment by markstory** [comment]
I don't know much about postgres to be honest, especially not the unixsocket approach, as I generally always just use TCP connections. It sounds like you know more though, perhaps you could do a pull request to update the documentation?

**Comment by bhoffman13** [comment]
@markstory I will attempt to help update the docs sometime this week.

**Comment by markstory** [comment]
@bhoffman13 Awesome!
